pulp_ansible/app/tasks/signature.py

```diff
@@ -0,0 +1,91 @@
+import asyncio
+import os
+import json
+import tempfile
+
+from pulpcore.plugin.stages import (
+    ArtifactSaver,
+    ContentSaver,
+    DeclarativeArtifact,
+    DeclarativeContent,
+    DeclarativeVersion,
+    Stage,
+    QueryExistingArtifacts,
+    QueryExistingContents,
+)
+
+from pulp_ansible.app.models import (
+    AnsibleRepository,
+    CollectionVersion,
+    CollectionVersionSignature,
+)
+
+from pulpcore.plugin.models import Artifact, SigningService
+from pulpcore.plugin.sync import sync_to_async_iterable
+
+
+def sign(repository_href, content_hrefs, signing_service_href):
+    """The signing task."""
+    repository = AnsibleRepository.objects.get(pk=repository_href)
+    if content_hrefs == ["*"]:
+        content = repository.latest_version().content.filter(pulp_type=CollectionVersion.get_pulp_type())
+    else:
+        content = CollectionVersion.objects.filter(pk__in=content_hrefs)
+    signing_service = SigningService.objects.get(pk=signing_service_href)
+    first_stage = CollectionSigningStage(content, signing_service)
+    SigningDeclarativeVersion(first_stage, repository).create()
+
+
+class SigningDeclarativeVersion(DeclarativeVersion):
+    """Custom signature pipeline."""
+
+    def pipeline_stages(self, new_version):
+        """The stages for the signing process."""
+        pipeline = [
+            self.first_stage,  # CollectionSigningStage
+            QueryExistingArtifacts(),
+            ArtifactSaver(),
+            # QueryExistingContents(),
+            ContentSaver(),
+        ]
+        return pipeline
+
+
+class CollectionSigningStage(Stage):
+    """
+    This stage signs the content and creates CollectionVersionSignatures.
+    """
+
+    def __init__(self, content, signing_service):
+        super().__init__()
+        self.content = content
+        self.signing_service = signing_service
+        self.base_dir = os.getcwd()
+
+    async def run(self):
+        """does the magic."""
+        async for collection_version in sync_to_async_iterable(self.content.iterator()):
+            # Can add extra parallelization here with tasks
```

Comment:
If you mean task in the tasking system, we avoid keeping "this task running and dispatching another that we need to wait on" because in cases where the worker count is 1 it would deadlock. For parallelization I think we can get coroutine based parallelization with the design in my comment below (maybe).

Reply:
This comment is referring to async tasks.